core/perf_test/test_reduction.cpp

```diff
@@ -0,0 +1,121 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <benchmark/benchmark.h>
+#include "Benchmark_Context.hpp"
+
+// Returns time for reduction in us
+template <class F, class R>
+void run_reduction(int N, F f, R& r) {
+  Kokkos::parallel_reduce(N, f, r);
+  // fence because r could be a View and thus
+  // parallel_reduce can be asyncronous
+  Kokkos::fence();
+}
+
+template <class T, size_t TestCase>
+static void Test_Reduction(benchmark::State& state) {
+  int N = state.range(0);
+
+  // Prep some Views used for benchmark
+  // avoids allocating/deallocating during benchmark phase
+  Kokkos::View<T*> data("data", N);
+  Kokkos::View<T> d_result("d_result");
+  Kokkos::View<T, Kokkos::HostSpace> h_result("h_result");
+  Kokkos::parallel_for(
+      "FillData", N, KOKKOS_LAMBDA(int i) { data(i) = i; });
+
+  // Simple kernel with little work
+  auto f_simple = KOKKOS_LAMBDA(int i, T& val) { val += data[i]; };
+
+  // Expensive kernel with more work per thread
+  auto f_expensive = KOKKOS_LAMBDA(int i, T& val) {
+    T result = data(i);
+    for (int k = 0; k < 100 + static_cast<int>(data(i)) % 2; k++) {
```

Comment:
Why `static_cast<int>(data(i)) % 2` and not just `i%2`?
Also what is the goal here?  Have some threads that do 100 iterations, others with 101 and 102?
I don't get it.